File: src/lib/cloud/cost.js
// Rough monthly cost estimator for Ground0: Cloud — Architecture Studio.
//
// estimateCost({ nodes }, { rps, instanceCounts }) -> { items, total }
//   rps            : request rate to scale per-request charges (0 = provisioned only)
//   instanceCounts : optional { nodeId: n } override from the load test (scaled fleet)
//
// Illustrative only — see specs.js. The point is to show beginners that design
// choices have a price, and that scaling up costs money.
import {
  HOURS_PER_MONTH, SECONDS_PER_MONTH, PRICING,
  instanceHourly, dbHourly, cacheHourly,
} from './specs.js'

function nodeMonthly(node, rps, instances) {
  const reqM = (rps * SECONDS_PER_MONTH) / 1e6 // millions of requests / month
  const c = node.config || {}
  switch (node.kind) {
    case 'compute':
    case 'container': {
      const monthly = instances * instanceHourly(c.instanceType) * HOURS_PER_MONTH
      return { monthly, detail: `${instances}× ${c.instanceType || 'instance'}` }
    }
    case 'serverless':
      return { monthly: reqM * PRICING.lambdaPerMillionReq, detail: 'pay per invocation' }
    case 'database': {
      const replicas = c.readReplicas || 0
      const azMult = c.multiAz ? 2 : 1
      const monthly = dbHourly(c.dbClass) * HOURS_PER_MONTH * (azMult + replicas)
      const parts = [c.dbClass || 'db']
      if (c.multiAz) parts.push('Multi-AZ')
      if (replicas) parts.push(`${replicas} replica${replicas > 1 ? 's' : ''}`)
      return { monthly, detail: parts.join(' · ') }
    }
    case 'cache':
      return { monthly: cacheHourly(c.cacheType) * HOURS_PER_MONTH, detail: c.cacheType || 'cache node' }
    case 'lb':
      return { monthly: PRICING.albMonthly + reqM * PRICING.albPerMillionReq, detail: 'ALB + LCUs' }
    case 'gateway':
      return { monthly: reqM * PRICING.apigwPerMillionReq, detail: 'per million requests' }
    case 'cdn':
      return { monthly: reqM * PRICING.cdnPerMillionReq, detail: 'requests + transfer' }
    case 'client':
      return { monthly: 10 + reqM * 0.4, detail: 'frontend hosting + requests' }
    case 'waf':
      return { monthly: PRICING.wafMonthly + reqM * PRICING.wafPerMillionReq, detail: 'WAF + requests' }
    case 'nat':
      return { monthly: PRICING.natMonthly + reqM * PRICING.natPerMillionReq, detail: 'gateway + data' }
    case 'dns':
      return { monthly: PRICING.dnsMonthly + reqM * PRICING.dnsPerMillionReq, detail: 'hosted zone + queries' }
    case 'storage':
      return { monthly: PRICING.s3Monthly + reqM * PRICING.s3PerMillionReq, detail: 'storage + requests' }
    case 'queue':
      return { monthly: reqM * PRICING.sqsPerMillionReq, detail: 'per million messages' }
    default:
      return { monthly: 0, detail: 'no charge' }
  }
}

export function estimateCost({ nodes = [] }, { rps = 0, instanceCounts = {} } = {}) {
  const items = nodes
    .filter((n) => n.kind !== 'internet')
    .map((n) => {
      const instances = instanceCounts[n.id] ?? n.config?.instances ?? 1
      const { monthly, detail } = nodeMonthly(n, rps, instances)
      return { id: n.id, name: n.label || n.name, icon: n.icon, kind: n.kind, monthly, detail }
    })
    .sort((a, b) => b.monthly - a.monthly)
  const total = items.reduce((s, i) => s + i.monthly, 0)
  return { items, total }
}

File: go-server/internal/studio/cloud/cost.go
// cost.go ports src/lib/cloud/cost.js's estimateCost. Illustrative-only
// monthly cost estimate; see specs_repo.go for where the pricing numbers
// come from.
package cloud

import (
	"context"
	"fmt"
	"sort"

	"ground0.dev/goserver/internal/model"
)

type CostItem struct {
	ID      string  `json:"id"`
	Name    string  `json:"name"`
	Icon    string  `json:"icon,omitempty"`
	Kind    string  `json:"kind"`
	Monthly float64 `json:"monthly"`
	Detail  string  `json:"detail"`
}

type CostResult struct {
	Items []CostItem `json:"items"`
	Total float64    `json:"total"`
}

func nodeMonthly(specs specsSnapshot, node model.Node, rps float64, instances float64) (float64, string) {
	reqM := rps * specs.constant("secondsPerMonth") / 1e6
	hoursPerMonth := specs.constant("hoursPerMonth")
	c := node.Config

	switch node.Kind {
	case "compute", "container":
		instanceType := configString(c, "instanceType")
		monthly := instances * specs.instanceHourly(instanceType) * hoursPerMonth
		label := instanceType
		if label == "" {
			label = "instance"
		}
		return monthly, fmt.Sprintf("%d× %s", int(instances), label)
	case "serverless":
		return reqM * specs.constant("lambdaPerMillionReq"), "pay per invocation"
	case "database":
		replicas, _ := configNumber(c, "readReplicas")
		multiAz, _ := configBool(c, "multiAz")
		azMult := 1.0
		if multiAz {
			azMult = 2
		}
		dbClass := configString(c, "dbClass")
		monthly := specs.dbHourly(dbClass) * hoursPerMonth * (azMult + replicas)
		parts := dbClass
		if parts == "" {
			parts = "db"
		}
		if multiAz {
			parts += " · Multi-AZ"
		}
		if replicas > 0 {
			suffix := "replica"
			if replicas > 1 {
				suffix = "replicas"
			}
			parts += fmt.Sprintf(" · %d %s", int(replicas), suffix)
		}
		return monthly, parts
	case "cache":
		cacheType := configString(c, "cacheType")
		label := cacheType
		if label == "" {
			label = "cache node"
		}
		return specs.cacheHourly(cacheType) * hoursPerMonth, label
	case "lb":
		return specs.constant("albMonthly") + reqM*specs.constant("albPerMillionReq"), "ALB + LCUs"
	case "gateway":
		return reqM * specs.constant("apigwPerMillionReq"), "per million requests"
	case "cdn":
		return reqM * specs.constant("cdnPerMillionReq"), "requests + transfer"
	case "client":
		return 10 + reqM*0.4, "frontend hosting + requests"
	case "waf":
		return specs.constant("wafMonthly") + reqM*specs.constant("wafPerMillionReq"), "WAF + requests"
	case "nat":
		return specs.constant("natMonthly") + reqM*specs.constant("natPerMillionReq"), "gateway + data"
	case "dns":
		return specs.constant("dnsMonthly") + reqM*specs.constant("dnsPerMillionReq"), "hosted zone + queries"
	case "storage":
		return specs.constant("s3Monthly") + reqM*specs.constant("s3PerMillionReq"), "storage + requests"
	case "queue":
		return reqM * specs.constant("sqsPerMillionReq"), "per million messages"
	default:
		return 0, "no charge"
	}
}

// EstimateCost ports src/lib/cloud/cost.js's estimateCost.
func EstimateCost(ctx context.Context, repo *SpecsRepo, nodes []model.Node, rps float64, instanceCounts map[string]int) CostResult {
	specs := loadSpecsSnapshot(ctx, repo)
	items := []CostItem{}
	for _, n := range nodes {
		if n.Kind == "internet" {
			continue
		}
		instances := 1.0
		if c, ok := instanceCounts[n.ID]; ok {
			instances = float64(c)
		} else if v, ok := configNumber(n.Config, "instances"); ok {
			instances = v
		}
		monthly, detail := nodeMonthly(specs, n, rps, instances)
		items = append(items, CostItem{ID: n.ID, Name: n.DisplayName(), Icon: n.Icon, Kind: n.Kind, Monthly: monthly, Detail: detail})
	}
	sort.Slice(items, func(i, j int) bool { return items[i].Monthly > items[j].Monthly })
	total := 0.0
	for _, it := range items {
		total += it.Monthly
	}
	return CostResult{Items: items, Total: total}
}
